app/controllers/api/v1/tags_controller.rb

```diff
@@ -0,0 +1,67 @@
+# frozen_string_literal: true
+
+module API
+  module V1
+    # Controller class that is responsible for handling tag requests.
+    class TagsController < ActionController::API
+      include DeviseTokenAuth::Concerns::SetUserByToken
+      include Pundit
+
+      before_action :authenticate_user!
+      before_action :set_tag, only: %i[show profiles posts]
+      before_action :authorize_tag!
+      after_action :verify_authorized
+
+      # GET /tags or /tags.json
+      def index
+        @tag = Tag.all
+        render json: @tag, status: :ok
+      end
+
+      # GET /tags/:id or /tags/:id.json
+      def show
+        render json: @tag, status: :ok
+      end
+
+      # POST /tags or /tags.json
+      def create
+        if tag_params.blank?
+          render json: { error: 'There was no tag data passed in so your tag could not be saved.' }, status: :unprocessable_entity
+        else
+          @tag = Tag.new(tag_params)
+          if @tag.save
+            render json: @tag, status: :created
+          else
+            render json: @tag.errors, status: :unprocessable_entity
+          end
+        end
+      end
+
+      # GET /tags/:tag_id/profiles or /tags/:tag_id/profiles.json
+      def profiles
+        profiles = @tag.profiles
+        render json: profiles, status: :ok
+      end
+
+      # GET /tags/:tag_id/posts or /tags/:tag_id/posts.json
+      def posts
+        posts = @tag.posts
```

Comment:
does it return public posts only or private + public posts?

https://apidock.com/rails/ActiveRecord/Base/default_scope/class
The easy fix for that is to add `default_scope` and set it to `public`.
Don't forget to add scopes `hidden` and `public` to `Profile` model